src/src/PluginStructs/P087_data_struct.cpp

```diff
@@ -39,11 +41,66 @@ bool P087_data_struct::init(ESPEasySerialPort port, const int16_t serial_rx, con
 }
 
 void P087_data_struct::post_init() {
+  String log = F("P087_data_init:");
+  hex_head_match   = 0;
+  hex_head_match_a = false;
+  hex_head_match_b = false;
+  sentence_part    = String();
+  hex_head_empty   = _lines[P087_HEX_HEADER_POS].isEmpty();
+  hex_data_length  = _lines[P087_HEX_DATA_LEN_POS].toInt();
+  log += F(" HEX Data Length: ");
+  log += String(hex_data_length);
+  if (hex_data_length > 0) {
+    hex_data_length  = _lines[P087_HEX_DATA_LEN_POS].toInt() * 2;
+    max_length       = 2048;
+  }
+  if (hex_data_length < 0) {
+    hex_data_add_length  = _lines[P087_HEX_DATA_LEN_ADD_POS].toInt();
+    log += F(" HEX Data Adding Length: ");
+    log += String(hex_data_add_length);
+    hex_data_add_length  = _lines[P087_HEX_DATA_LEN_ADD_POS].toInt() * 2;
+    max_length       = 2048;
+  }
+  if (hex_data_length != 0 && !hex_head_empty) {
+    String buf = _lines[P087_HEX_HEADER_POS];
+    log += F(" HEX Header: ");
+    log += buf;
+    uint16_t len = buf.length();
+    head_sentence_length = (len / 2);
+    log += F(" HEX Header length: ");
+    log += String(head_sentence_length);
+    for(int i = 0; i < len / 2 && !hex_head_empty; i++) {
+      uint8_t uc = buf[i * 2], dc = buf[i * 2 + 1];
+      if(uc >= 48 && uc <= 57) {
+        uc -= 48;
+      } else if(uc >= 65 && uc <= 70) {
+        uc -= 55;
+      } else if(uc >= 97 && uc <= 102) {
+        uc -= 87;
+      } else {
+        hex_head_empty = true;
+      }
+
+      if(dc >= 48 && dc <= 57) {
+        dc -= 48;
+      } else if(dc >= 65 && dc <= 70) {
+        dc -= 55;
+      } else if(dc >= 97 && dc <= 102) {
+        dc -= 87;
+      } else {
+        hex_head_empty = true;
+      }
+      if(!hex_head_empty) {
+        head_sentence[i] = (uc << 4 | dc);
+      }
+    }
+  }
```

Comment:
This is repeated quite a few times in your code, better have a hex decoder function for it.